### DISCUSSION TITLE
Dynamodb table schema

### DISCUSSION THREAD

**Discussion by Shir-Avr** [reclosedev/requests-cache#691] [Q&A] [answered]
Hi,

Looking at the docs of the DynamoDbCache object [(link to DynamoDbCache docs)](https://requests-cache.readthedocs.io/en/stable/modules/requests_cache.backends.dynamodb.html),
it says the partition key for the dynamodb table is "namespace".
In the constructor's params, the namespace is described as "Name of DynamoDB hash map" 

Going over at the dynamodb table itself - all records have the same partition key "responses" (see attached image)
**My question is why all records have the same partition key - "responses"?**

I feel like it missed the whole point of using dynamodb (which is a fully managed database)
Since by using it this way, the table uses only one partition and all records are written to the same partition.
 
 Why doesn't it use the cache key as the partition key?
 thanks!

<img width="775" alt="Screen Shot 2022-09-19 at 16 32 07" src="https://user-images.githubusercontent.com/86551477/191035902-c0c40b60-3b43-47d9-ad79-deb527a5de57.png">
<img width="744" alt="Screen Shot 2022-09-19 at 16 32 15" src="https://user-images.githubusercontent.com/86551477/191029204-0d153041-398d-4a93-a0f6-d64da19d33a6.png">
<img width="694" alt="Screen Shot 2022-09-19 at 16 39 04" src="https://user-images.githubusercontent.com/86551477/191030596-a22a7c83-0cf0-4916-ac24-31877a451d97.png">

@JWCook

**Comment by JWCook** [comment, marked as answer]
Thanks for bringing this up. The short answer is that you're correct; the cache key is currently used as the range key, but would be better as the partition key.

The longer answer is that there is a second namespace, `redirects`, which optimizes redirect chains by mapping the cache key of the initial request to the cache key of the final request in the chain. Other backends use a separate table to store this redirect info, but the DynamoDB backend uses a single table with two different "namespaces" (partition key values). That was done [in the initial implementation here](https://github.com/requests-cache/requests-cache/pull/92) before I came on board as maintainer, and I did not want to break backwards-compatibility.

However, since the [1.0 release is coming soon](https://github.com/requests-cache/requests-cache/issues/402), this would be as good a time as any to make some backwards-incompatible improvements. There are a couple different options, but probably the easiest would be to use two separate tables, remove `namespace`, and make `key` the partition key.

What do you think? And is that something you would be interested in working on?

**Reply by JWCook** [reply]
@Shir-Avr These changes have been made in #795.

One compromise is that the DynamoDB backend no longer does persistent redirect optimization, and now just stores this info in-memory. Unlike the other backends, with DynamoDB I think the situational performance boost from this is not worth the extra read throughput needed, plus using a separate table or index to store redirects.
